compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEAttrs.cpp

```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEAttrs.h"
+
+#include "iree-amd-aie/IR/AMDAIEDialect.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Transform/IR/TransformOps.h"
+#include "mlir/IR/DialectImplementation.h"
+
+#define GET_ATTRDEF_CLASSES
+#include "iree-amd-aie/IR/AMDAIEAttrs.cpp.inc"
+#include "iree-amd-aie/IR/AMDAIEEnums.cpp.inc"
+
+static const char kPackingConfigAttrName[] = "packing_config";
+
+namespace mlir::iree_compiler {
+
+/// Returns an `ArrayAttr` where each element is an `IntegerAttr` of 64-bit
+/// integer type whose values is obtained from `values`.
+static ArrayAttr getIndexArrayAttr(MLIRContext *context,
+                                   ArrayRef<int64_t> values) {
+  return ArrayAttr::get(
+      context, llvm::map_to_vector(values, [&](int64_t value) -> Attribute {
+        return IntegerAttr::get(IndexType::get(context), APInt(64, value));
+      }));
+}
+
+}  // namespace mlir::iree_compiler
+
+namespace mlir::iree_compiler::AMDAIE {
+
+//===----------------------------------------------------------------------===//
+// amdaie.packing_config
+//===----------------------------------------------------------------------===//
+
+PackingConfigAttr PackingConfigAttr::get(
+    MLIRContext *context, PackingConfigListTypeRef packingConfigs) {
+  Builder builder(context);
+  SmallVector<PackingConfigPackingLevelAttr> packinglevels;
+  for (auto [level, configs] : llvm::enumerate(packingConfigs)) {
+    // Form `innerPerm` attribute which is a multi-dimensional array attribute.
+    SmallVector<PermLevelAttr> innerPermLevels;
+    for (auto row : configs.innerPerm) {
+      innerPermLevels.push_back(PermLevelAttr::get(context, row));
+    }
+    auto innerPerm = PermLevelsAttr::get(context, innerPermLevels);
+
+    // Form `outerPerm` attribute which is a multi-dimensional array attribute.
+    SmallVector<PermLevelAttr> outerPermLevels;
+    for (auto row : configs.outerPerm) {
+      outerPermLevels.push_back(PermLevelAttr::get(context, row));
+    }
+    auto outerPerm = PermLevelsAttr::get(context, outerPermLevels);
+
+    packinglevels.push_back(PackingConfigPackingLevelAttr::get(
+        context, configs.packedSizes, configs.transposePackIndices,
+        configs.unpackEmpty, innerPerm, outerPerm));
+  }
+  return get(context,
+             PackingConfigPackingLevelsAttr::get(context, packinglevels));
+}
+
+PackConfig PackingConfigAttr::getPackingConfigVals(unsigned level) {
+  Builder builder(getContext());
+  auto levels = getPackingLevels();
+  if (level >= levels.size()) return {};
+  PackConfig packConfig;
+  SmallVector<OpFoldResult> packedSizes;
+  for (int64_t packedSize : levels[level].getPackedSizes()) {
+    packedSizes.push_back(builder.getI64IntegerAttr(packedSize));
+  }
+  packConfig.packedSizes = packedSizes;
```

Comment:
This is the consequence of the stopgap solution that I mentioned above.

My vote for keeping `PackingConfigTD` will be for the "cleanest" representation it provides without involving any of `i32`/`i64`/`tensor<axbx...xi32>`.

My vote for keeping `PackingConfig` will be because it seamlessly integrates with the `iree-amdaie-pack-and-transpose`.

A solution that I propose :-
API_1. Have one API which would just extract `packedSizes` and return the `SmallVector<OpFoldResult>`.
API_2: Have one API which would extract level based config in `ArrayRef` representation.

Adapt `iree-amdaie-pack-and-transpose` to use API_1 for packing and API_2 for transpose (ignoring the `packedSizes`  in case of the latter since it anyway doesn't seem to be using that during transpose).

Let me know your thoughts.